torch/optim/lr_scheduler.py

```diff
@@ -427,3 +427,83 @@ def state_dict(self):
     def load_state_dict(self, state_dict):
         self.__dict__.update(state_dict)
         self._init_is_better(mode=self.mode, threshold=self.threshold, threshold_mode=self.threshold_mode)
+
+
+class SGDR(_LRScheduler):
```

Comment:
This scheduler is not tied to SGD but can also be effectively used for other optimizers such as Adam, so although it's in the name of the paper, I find the name `SGDR` misleading. Why not call it `CosineAnnealingWR` or `CosineAnnealingWarmRestarts`?

Reply:
you are right, that's fine with me
What do you think, @mrshenli ?

Reply:
Sounds good to me too. 

Curious, there is an `CosineAnnealingLR` which implements the SGDR without warm restarts. Do you know what does LR mean here? (learning rate? less restart?) I just want to make sure if we go with `CosineAnnealingWR`, the naming does not confuse people.

Reply:
CosineAnnealingWarmRestarts sounds better to me.

Reply:
hi, @mrshenli 
It's a kind of learning rate policy acoording to [Cyclical Learning Rates for Training Neural Networks](https://arxiv.org/pdf/1506.01186.pdf). So, in my opion, it's learning rate.

Reply:
Hi, @mrshenli 
how about I push a version of integer `T_i`, I will test the CI error after #18244 is done?

Reply:
That sounds good to me. Please edit the doc and arg check accordingly. I will approve and merge. Thanks!

Reply:
Hi, @mrshenli 
There two ways to convert `T_i` to an integer. Because `T_i` equals to `T_{i - 1} * T_mult`, we can use `int(T_mult)` or `int(T_{i - 1} * T_mult)` to force `T_i` to be an integer. However, the latter would case a inconsistency, for example, let `T_0 = 10`, `T_mult = 1.5`, if we call `scheduler.step()` for 100 times, and call `scheduler.step(100)` the `T_i` would be different.
The `T_i` for the example would be `[10, 15, 22, 33, 49]` and 50.625 (50 after int()) for the formula result. So for simplicity, I choose the former way.